plugins/genesys/fu-genesys-usbhub-firmware.c

```diff
@@ -240,197 +298,17 @@ fu_genesys_usbhub_firmware_parse(FuFirmware *firmware,
-					  sizeof(self->static_ts.running_project_code));
-	fu_xmlb_builder_insert_kv(bn, "running_project_code", running_project_code);
-
-	running_project_hardware =
-	    fu_strsafe((const gchar *)&self->static_ts.running_project_hardware,
-		       sizeof(self->static_ts.running_project_hardware));
-	if (running_project_hardware != NULL)
-		running_project_hardware[0] += 0x10; /* '1' -> 'A'... */
-	fu_xmlb_builder_insert_kv(bn, "running_project_hardware", running_project_hardware);
-
-	running_project_firmware =
-	    fu_strsafe((const gchar *)&self->static_ts.running_project_firmware,
-		       sizeof(self->static_ts.running_project_firmware));
-	fu_xmlb_builder_insert_kv(bn, "running_project_firmware", running_project_firmware);
-
-	running_project_ic_type = fu_genesys_usbhub_firmware_get_project_ic_type_string(
-	    (const gchar *)self->static_ts.running_project_ic_type);
-	fu_xmlb_builder_insert_kv(bn, "running_project_ic_type", running_project_ic_type);
-}
-
-static gboolean
-fu_genesys_usbhub_firmware_build(FuFirmware *firmware, XbNode *n, GError **error)
-{
-	FuGenesysUsbhubFirmware *self = FU_GENESYS_USBHUB_FIRMWARE(firmware);
-	const gchar *tmp;
-	guint64 tmp64;
-
-	/* optional properties */
-	tmp64 = xb_node_query_text_as_uint(n, "tool_string_version", NULL);
-	if (tmp64 != G_MAXUINT64) {
-		if (tmp64 > G_MAXUINT8) {
-			g_set_error_literal(error,
-					    G_IO_ERROR,
-					    G_IO_ERROR_INVALID_DATA,
-					    "invalid tool_string_version");
-			return FALSE;
-		}
-		self->static_ts.tool_string_version = tmp64;
-	}
+	g_autofree gchar *static_ts = NULL;
 
-	/* mask_project_code */
-	tmp = xb_node_query_text(n, "mask_project_code", NULL);
-	if (tmp != NULL) {
-		gsize len = strlen(tmp);
-		if (len != 4) {
-			g_set_error(error,
-				    G_IO_ERROR,
-				    G_IO_ERROR_INVALID_DATA,
-				    "invalid mask_project_code %s, got 0x%x length",
-				    tmp,
-				    (guint)len);
-			return FALSE;
-		}
-		if (!fu_memcpy_safe((guint8 *)&self->static_ts.mask_project_code,
-				    sizeof(self->static_ts.mask_project_code),
-				    0x0, /* dst */
-				    (const guint8 *)tmp,
-				    len,
-				    0x0, /* src */
-				    len,
-				    error))
-			return FALSE;
-	}
-
-	/* mask_project_ic_type */
-	tmp = xb_node_query_text(n, "mask_project_ic_type", NULL);
-	if (tmp != NULL) {
-		gsize len = strlen(tmp);
-		if (len != 6) {
-			g_set_error(error,
-				    G_IO_ERROR,
-				    G_IO_ERROR_INVALID_DATA,
-				    "invalid mask_project_ic_type %s, got 0x%x length",
-				    tmp,
-				    (guint)len);
-			return FALSE;
-		}
-		if (!fu_memcpy_safe((guint8 *)&self->static_ts.mask_project_ic_type,
-				    sizeof(self->static_ts.mask_project_ic_type),
-				    0x0, /* dst */
-				    (const guint8 *)tmp,
-				    len,
-				    0x0, /* src */
-				    len,
-				    error))
-			return FALSE;
-	}
-
-	/* success */
-	return TRUE;
+	static_ts = fu_strsafe((const gchar *)self->fu_struct_static_ts->data,
```

Comment:
Could this be defined as a string m then you get the helper for free.

Reply:
I cannot understand. May I ask for a clearer explanation?

Reply:
Ohh, I'm a bit confused about what you're trying to do -- isn't self->fu_struct_static_ts->data just binary data?

I think it would be much more helpful to use fu_xmlb_builder_insert_kv() for each of fu_struct_genesys_ts_static_get_mask_project_code(), fu_struct_genesys_ts_static_get_firmware_version() etc.

Reply:
The log as shown below with `--daemon-verbose=FuStruct --daemon-verbose=FuPluginGenesys`.
```
09:11:30.798 FuStruct             GenesysTsStatic:
  tool_string_version: 0x37 [dynamic-13-byte]
  mask_project_code: 7455
  mask_project_hardware: 1
  mask_project_firmware: 03
  mask_project_ic_type: 352510
  running_project_code: 7455
  running_project_hardware: 1
  running_project_firmware: 03
  running_project_ic_type: 352510
  firmware_version: 8901

<firmware gtype="FuGenesysUsbhubFirmware">
  <flags>has-checksum</flags>
  <version>89.01</version>
  <version_raw>0x8901</version_raw>
  <size>0xb000</size>
  <data size="0x20000">.U~.-.GLHUB.he..</data>
  <static_ts>7745510335251074551033525108901</static_ts>
</firmware>
```